Add <CommandPalette> (#799)
universal-ember/ember-primitives: docs-app/app/templates/5-floaty-bits/command-palette.gjs.md

Reviewer: we should have a demo that simulates searching using the starwars api (swapi.dev, I think)

# Command Palette

A search box over a list of results, in a modal `<dialog>`. The <kbd>⌘</kbd><kbd>K</kbd> pattern.

`<CommandPalette>` gives you the input, the keyboard, and the aria wiring. It does not filter. You render the results you want, in the order you want, from an array, a fetch, or a ranked search index.

<div class="featured-demo">

> ```gjs live preview no-shadow

Reviewer: we need demos for different situations:
- multiline descriptions
- not in a modal

  { icon: '⚙️', label: 'Open Settings', hint: 'Preferences and keybindings' },
  { icon: '🌓', label: 'Toggle Theme', hint: 'Switch between light and dark' },
];

const matching = (query) =>
  COMMANDS.filter((command) =>
    command.label.toLowerCase().includes(query.toLowerCase().trim())
  );

const announce = (command) => alert(`Ran: ${command.label}`);

<template>
  <CommandPalette @onSelect={{announce}} @hotkey="mod+k" as |p|>
    <p.Trigger class="trigger">
      Search commands
      <kbd>⌘K</kbd>
    </p.Trigger>

    <p.Dialog class="palette">
      <p.Input class="query" placeholder="Type a command…" aria-label="Command" />

      <p.List class="results" as |l|>
        {{#each (matching p.query) as |command|}}
          <l.Item class="result" @value={{command}}>
            <span class="icon">{{command.icon}}</span>
            <span>
              <strong>{{command.label}}</strong>
              <small>{{command.hint}}</small>
            </span>
          </l.Item>
        {{else}}
          <p class="empty">No commands match “{{p.query}}”.</p>
        {{/each}}
      </p.List>
    </p.Dialog>
  </CommandPalette>

>   <style>

Reviewer: css is way off here. visually verify

>     .trigger {

Reviewer: use `@scope`

      color: light-dark(#18181b, #fafafa);
      box-shadow: 0 25px 50px -12px rgb(0 0 0 / 0.35);
    }

    .palette::backdrop {
      background: rgb(0 0 0 / 0.4);
      backdrop-filter: blur(2px);
    }

    .query {
      width: 100%;
      padding: 1rem;
      border: none;
      border-block-end: 1px solid light-dark(#e4e4e7, #27272a);
      background: none;
      color: inherit;
      font-size: 1rem;
    }

    .query:focus-visible {
      outline: none;
    }

    .results {
      max-height: 18rem;
      overflow-y: auto;
      padding: 0.5rem;
    }

    .result {
      display: flex;
      gap: 0.75rem;
      align-items: center;
      padding: 0.5rem 0.75rem;
      border-radius: 8px;
      cursor: pointer;
    }

    /* the active option -- what Enter would choose */
    .result[data-active='true'] {
      background: light-dark(#f4f4f5, #27272a);
    }

    .result small {
      display: block;
      opacity: 0.6;
    }

    .empty {
      padding: 1.5rem;
      text-align: center;
      opacity: 0.6;
    }
  </style>
</template>
```

</div>

Press <kbd>Ctrl</kbd> / <kbd>⌘</kbd> + <kbd>K</kbd> anywhere on this page to open it.

> ## How it is put together

Reviewer: this isnt relevant for readers

# Command Palette

A combobox over a listbox. The <kbd>⌘</kbd><kbd>K</kbd> pattern.

`<CommandPalette>` gives you the input, the keyboard, and the aria wiring. It does not filter, and it does not open or close anything. Put it inside a [`<Dialog>`](/5-floaty-bits/modal.md) for the <kbd>⌘</kbd><kbd>K</kbd> case, or render it on a page of its own.

<div class="featured-demo">

```gjs live preview no-shadow
import { CommandPalette, Dialog } from "ember-primitives";

> import { cell } from "ember-resources";

Reviewer: use tracked(""), not cell

const lastRun = cell("");
const run = (label) => () => lastRun.set(label);

const COMMANDS = [
  { icon: "📄", label: "New File", description: "Create a file in this folder", onSelect: run("New File") },
  { icon: "📁", label: "New Folder", description: "Create a folder", onSelect: run("New Folder") },
  { icon: "🔍", label: "Find in Files", description: "Search the whole project", onSelect: run("Find in Files") },
  { icon: "⚙️", label: "Open Settings", description: "Preferences and keybindings", onSelect: run("Open Settings") },
  { icon: "🌓", label: "Toggle Theme", description: "Switch between light and dark", onSelect: run("Toggle Theme") },
];

<template>

>   <Dialog class="palette" closedby="any" as |d|>

Reviewer: you don't need a class here -- closedby shouldn't be specified either -- it should be the best default possible. also, Dialog will need its own docs and tests.

>   <Dialog class="palette" closedby="any" as |d|>

Reviewer: you don't need a class here -- closedby shouldn't be specified either -- it should be the best default possible. also, Dialog will need its own docs and tests.

      }

      :scope::backdrop {
        background: rgb(0 0 0 / 0.45);
      }

      input {
        width: 100%;
        padding: 1rem;
        border: none;
        border-block-end: 1px solid light-dark(#e4e4e7, #27272a);
        background: none;
        color: inherit;
        font-size: 1rem;
      }

      input:focus-visible {
        outline: none;
      }

      [role="listbox"] {
        max-height: 18rem;
        overflow-y: auto;
        padding: 0.5rem;
      }

      [role="option"] {
        display: grid;
        grid-template-columns: 1.5rem 1fr;
        gap: 0 0.75rem;
        padding: 0.5rem 0.75rem;
        border-radius: 8px;
        cursor: pointer;
      }

      /* the pointer and the keyboard set the same one. Do not use :hover */
      [role="option"][data-active="true"] {
        background: light-dark(#f4f4f5, #27272a);
      }

      [role="option"] > :first-child {
        grid-row: span 2;
      }

      [role="option"] > :last-child {
        opacity: 0.6;
        font-size: 0.875rem;
      }
    }

    .last-run {
      margin-block-start: 0.75rem;
    }
  </style>
</template>
```

</div>

> ## Without a modal

Reviewer: the CSS on these demos is a bit messed up, you should debug.

        padding: 0.4rem;
      }

      [role="option"] {
        padding: 0.5rem 0.75rem;
        border-radius: 6px;
        cursor: pointer;
      }

      [role="option"][data-active="true"] {
        background: light-dark(#f4f4f5, #27272a);
      }

      .who {
        display: flex;
        gap: 0.5rem;
        align-items: baseline;
        margin: 0;
      }

      .who span {
        opacity: 0.6;
        font-size: 0.8rem;
      }

      /* a description that runs to several lines wraps rather than clipping */
      .note {
        margin: 0.15rem 0 0;
        opacity: 0.6;
        font-size: 0.875rem;
        line-height: 1.4;
      }

      .empty {
        padding: 1rem;
        text-align: center;
        opacity: 0.6;
      }
    }
  </style>
</template>
```

</div>

`LinkItem` is a row that is also a link. <kbd>Enter</kbd> dispatches a real click on the anchor, so the router navigates exactly as it would have for a mouse, and <kbd>⌘</kbd>-click still opens a new tab.

```hbs
<c.List as |l|>
  {{#each this.results as |result|}}
    <l.LinkItem @href={{result.path}}>{{result.title}}</l.LinkItem>
  {{/each}}
</c.List>
```

## Closing, and opening

> The palette does not open or close anything by itself. Hand it the two functions [`<Dialog>`](/5-floaty-bits/modal.md) yields:

Reviewer: let's give dialog its own docs page

      .note {
        margin: 0.15rem 0 0;
        color: var(--muted);
        font-size: 0.875rem;
        line-height: 1.4;
      }

      .empty {
        padding: 1rem;
        text-align: center;
        color: var(--muted);
      }
    }

  </style>
</template>
```

</div>

## Results that arrive later

Nothing about the rows has to be there when the reader types. Set them when they arrive, and the palette re-activates the best one, so <kbd>Enter</kbd> keeps meaning "the top result".

<div class="featured-demo auto-height">

```gjs live preview no-shadow
import { CommandPalette } from "ember-primitives";
import { tracked } from "@glimmer/tracking";

const people = tracked([]);
const status = tracked("Type a name: luke, skywalker, r2…");

/**
 * Responses can come back in a different order than the requests went out,
 * so a stale one is dropped rather than allowed to overwrite a newer one.
 */
let latest = 0;

const search = async (query) => {
  const mine = ++latest;
  const term = query.trim();

  if (term.length < 2) {
    people.value = [];
    status.value = "Type a name: luke, skywalker, r2…";
    return;
  }

  status.value = "Searching…";

  try {

>     const response = await fetch(

Reviewer: use the getPromiseState pattern -- don't manually manage status -- the promise is a root of state

Reviewer: this will allow us to have loading state, which this demo currently doesn't do well

      `https://swapi.dev/api/people/?search=${encodeURIComponent(term)}`
    );
    const json = await response.json();

    if (mine !== latest) return;

    people.value = json.results.map((person) => ({
      label: person.name,

>       description: `born ${person.birth_year} · ${person.height}cm · ${person.gender}`,

Reviewer: let's swap gender for their homeworld name

       * nesting takes from the most specific selector in its list.
       */
      :scope input[role="combobox"]:focus-visible {
        box-shadow: none;
        outline: 2px solid var(--ring);
        outline-offset: -2px;
        border-radius: 10px 10px 0 0;
      }

      [role="listbox"] {
        padding: 0.4rem;
      }

      [role="option"] {
        padding: 0.5rem 0.75rem;
        border-radius: 6px;
        cursor: pointer;
      }

      [role="option"][data-active="true"] {
        background: var(--active);
      }

      .who {
        display: flex;
        gap: 0.5rem;
        align-items: baseline;
        margin: 0;
      }

      .who span {
        color: var(--muted);
        font-size: 0.8rem;
      }

      /* a description that runs to several lines wraps rather than clipping */
      .note {
        margin: 0.15rem 0 0;
        color: var(--muted);
        font-size: 0.875rem;
        line-height: 1.4;
      }

      .empty {
        padding: 1rem;
        text-align: center;
        color: var(--muted);
      }
    }
  </style>
</template>
```

</div>

## Results that arrive later

> Nothing about the rows has to be there when the reader types. Set them when they arrive, and the palette re-activates the best one, so <kbd>Enter</kbd> keeps meaning "the top result".

Reviewer: rewrite these sentences. it's awk

       * element and so is clipped by a full-bleed input in a rounded box.
       * This one is drawn inside the input instead. Matching the role as
       * well as the element clears the site rule's specificity, which CSS
       * nesting takes from the most specific selector in its list.
       */
      :scope input[role="combobox"]:focus-visible {
        box-shadow: none;
        outline: 2px solid var(--ring);
        outline-offset: -2px;
        border-radius: 10px 10px 0 0;
      }

      [role="listbox"] {
        max-height: 14rem;
        overflow-y: auto;
        padding: 0.4rem;
      }

      [role="option"] {
        display: grid;
        padding: 0.5rem 0.75rem;
        border-radius: 6px;
        cursor: pointer;
      }

      [role="option"][data-active="true"] {
        background: var(--active);
      }

      [role="option"] > :last-child {
        color: var(--muted);
        font-size: 0.875rem;
      }

      .status {
        margin: 0;
        padding: 0.6rem 1rem;
        border-block-start: 1px solid var(--border);
        color: var(--muted);
        font-size: 0.8rem;
      }
    }
  </style>
</template>
```

</div>

> Rows arriving is the only thing that changed. `@onQueryChange` starts the request, `@items` takes whatever comes back, and the palette does the rest.

Reviewer: in this whole doc, don't use the word "arriving"

`LinkItem` is a row that is also a link. <kbd>Enter</kbd> dispatches a real click on the anchor, so the router navigates exactly as it would have for a mouse, and <kbd>⌘</kbd>-click still opens a new tab.

> `LinkItem` builds on [`<Link>`](/4-routing/link.md), which needs [`properLinks`](/4-routing/proper-links.md) set up in the application route for a plain `<a href>` to route rather than reload the page.

Reviewer: ```suggestion
`LinkItem` needs [`properLinks`](/4-routing/proper-links.md) set up in the application route for a plain `<a href>` support.
```

```hbs
<Dialog as |d|>
  <CommandPalette
    @items={{this.commands}}
    @onSelect={{d.close}}
    @onOpen={{d.open}}
    @hotkey="mod+k"
  />
</Dialog>
```

`<Dialog>` renders the element, so a trigger cannot sit outside it. That suits a palette, which opens on a key combination. When a button opens it instead, use [`<Modal>`](/5-floaty-bits/modal.md). The palette composes with either.

`@onSelect` runs every time a row is chosen, whichever row and however it was rendered. That is why `d.close` is all it takes to close on select. What a row does belongs to the row: `onSelect` on its `@items` entry, or `@onSelect` on its `Item`.

`@hotkey` calls `@onOpen`. Without `@onOpen` there is nothing to open, and no listener is installed.

## Styling

The active row carries `data-active="true"` and `aria-selected="true"`. Style either.

Do not use `:hover` for this. The pointer and the keyboard set the same active row, so one selector covers both, and the two can never disagree about what <kbd>Enter</kbd> will do.

## Install

```hbs live
<SetupInstructions @src="components/command-palette.gts" @since="0.62.0" />
```

> ## API Reference

Reviewer: note that a lot of information is misisng from here -- so we either need to update kolay, or we need to export (from just the command-palette file) the other referenced types

```gjs live no-shadow
import { ComponentSignature } from "kolay";

<template>
  <ComponentSignature
    @package="ember-primitives"
    @module="declarations/components/command-palette"
    @name="Signature"
  />
</template>
```

### Item

```gjs live no-shadow
import { ComponentSignature } from "kolay";

<template>
  <ComponentSignature
    @package="ember-primitives"
    @module="declarations/components/command-palette"
    @name="ItemSignature"
  />
</template>
```

> ## Accessibility

Reviewer: accessibility and keyboard interactions should go above the api reference section
